Full scan of huge number of images finally fails with "parse token error, token is expired by 3m5s"

When launching a full clair scan of all our images (> 25000), only the first images scanned in the first ~30min are processed. Then all remaining images scan fails with errors:
```
 2019/12/20 10:00:51 #033[1;44m[D] [server.go:2774] |    172.16.4.30|#033[42m 200 #033[0m|    2.87953ms|   match|#033[46m POST    #033[0m /service/notifications   r:/service/notifications#033[0m
 2019/12/20 10:00:51 #033[1;44m[D] [server.go:2774] |    172.16.4.29|#033[42m 200 #033[0m|  17.222827ms|   match|#033[46m POST    #033[0m /service/notifications/jobs/scan/fafd8c17-230a-11ea-8310-0242ac100433   r:/service/notifications/jobs/scan/:uuid#033[0m
 2019/12/20 10:00:51 #033[1;44m[D] [server.go:2774] |    172.16.4.31|#033[43m 401 #033[0m|  56.651688ms|   match|#033[44m GET     #033[0m /v2/xxxxx/manifests/sha256:96915dbcdfba2eed45dbd0694aef541af62f3408f73b1297437d3266a3xxxxxx   r:/v2/*#033[0m
 2019-12-20T10:00:51Z [ERROR] [/pkg/token/token.go:66]: parse token error, token is expired by 3m6s
 2019-12-20T10:00:51Z [ERROR] [/core/middlewares/regtoken/handler.go:51]: failed to decode reg token: token is expired by 3m6s, the error is skipped and round the request to native registry.
```

**Expected behavior and actual behavior:**
All images should be processed, if the token is expired it should be renewed for the process to continue

**Steps to reproduce the problem:**
* have many images in registry
* Go to Administration -> Interrogation Services -> Vulnerability
* click "scan"

**Versions:**
- harbor version: 1.10.0
- docker engine version: 18.09.2

--- a/src/common/config/context.go
+++ b/src/common/config/context.go
@@ -1,0 +1,32 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"context"
+)
+
+type cfgMgrKey struct{}
+
+// FromContext returns CfgManager from context
+func FromContext(ctx context.Context) (*CfgManager, bool) {
+	m, ok := ctx.Value(cfgMgrKey{}).(*CfgManager)
+	return m, ok
+}
+
+// NewContext returns context with CfgManager
+func NewContext(ctx context.Context, m *CfgManager) context.Context {
+	return context.WithValue(ctx, cfgMgrKey{}, m)
+}

--- a/src/jobservice/job/impl/context.go
+++ b/src/jobservice/job/impl/context.go
@@ -16,12 +16,12 @@ package impl
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"math"
 	"sync"
 	"time"
 
-	"errors"
 	comcfg "github.com/goharbor/harbor/src/common/config"
 	"github.com/goharbor/harbor/src/common/dao"
 	"github.com/goharbor/harbor/src/jobservice/config"
@@ -53,7 +53,7 @@ type Context struct {
 // NewContext ...
 func NewContext(sysCtx context.Context, cfgMgr *comcfg.CfgManager) *Context {
 	return &Context{
-		sysContext: sysCtx,
+		sysContext: comcfg.NewContext(sysCtx, cfgMgr),
 		cfgMgr:     *cfgMgr,
 		properties: make(map[string]interface{}),
 	}

--- a/src/pkg/robot/controller.go
+++ b/src/pkg/robot/controller.go
@@ -2,6 +2,8 @@ package robot
 
 import (
 	"fmt"
+	"time"
+
 	"github.com/dgrijalva/jwt-go"
 	"github.com/goharbor/harbor/src/common"
 	"github.com/goharbor/harbor/src/common/utils/log"
@@ -11,8 +13,9 @@ import (
 	"github.com/goharbor/harbor/src/pkg/token"
 	robot_claim "github.com/goharbor/harbor/src/pkg/token/claims/robot"
 	"github.com/pkg/errors"
-	"time"
 )
+
+// go:generate mockery -case underscore -name Controller
 
 var (
 	// RobotCtr is a global variable for the default robot account controller implementation

--- a/src/pkg/robot/model/robot.go
+++ b/src/pkg/robot/model/robot.go
@@ -1,11 +1,14 @@
 package model
 
 import (
+	"encoding/json"
+	"errors"
+	"time"
+
 	"github.com/astaxie/beego/orm"
 	"github.com/astaxie/beego/validation"
 	"github.com/goharbor/harbor/src/common/rbac"
 	"github.com/goharbor/harbor/src/common/utils"
-	"time"
 )
 
 // RobotTable is the name of table in DB that holds the robot object
@@ -32,6 +35,25 @@ type Robot struct {
 // TableName ...
 func (r *Robot) TableName() string {
 	return RobotTable
+}
+
+// FromJSON parses robot from json data
+func (r *Robot) FromJSON(jsonData string) error {
+	if len(jsonData) == 0 {
+		return errors.New("empty json data to parse")
+	}
+
+	return json.Unmarshal([]byte(jsonData), r)
+}
+
+// ToJSON marshals Robot to JSON data
+func (r *Robot) ToJSON() (string, error) {
+	data, err := json.Marshal(r)
+	if err != nil {
+		return "", err
+	}
+
+	return string(data), nil
 }
 
 // RobotQuery ...

--- a/src/pkg/scan/api/scan/base_controller.go
+++ b/src/pkg/scan/api/scan/base_controller.go
@@ -15,17 +15,13 @@
 package scan
 
 import (
-	"encoding/base64"
 	"fmt"
-	"time"
 
-	tk "github.com/docker/distribution/registry/auth/token"
 	cj "github.com/goharbor/harbor/src/common/job"
 	jm "github.com/goharbor/harbor/src/common/job/models"
 	"github.com/goharbor/harbor/src/common/rbac"
 	"github.com/goharbor/harbor/src/common/utils/log"
 	"github.com/goharbor/harbor/src/core/config"
-	"github.com/goharbor/harbor/src/core/service/token"
 	"github.com/goharbor/harbor/src/jobservice/job"
 	"github.com/goharbor/harbor/src/jobservice/logger"
 	"github.com/goharbor/harbor/src/pkg/robot"
@@ -228,9 +224,13 @@ func (bc *basicController) Scan(artifact *v1.Artifact, options ...Option) error 
 	}
 
 	// Get authorization type
-	auth := authorizationBasic
+	var auth string
 	if v, ok := meta.Properties[authorizationType]; ok {
 		auth = v
+	}
+
+	if auth != authorizationBasic && auth != authorizationBearer {
+		auth = authorizationBasic
 	}
 
 	jobID, err := bc.launchScanJob(trackID, artifact, r, producesMimes, auth)
@@ -346,14 +346,22 @@ func (bc *basicController) HandleJobHooks(trackID string, change *job.StatusChan
 	// Clear robot account
 	// Only when the job is successfully done!
 	if change.Status == job.SuccessStatus.String() {
-		if v, ok := change.Metadata.Parameters[sca.JobParameterRobotID]; ok {
-			if rid, y := v.(float64); y {
-				if err := robot.RobotCtr.DeleteRobotAccount(int64(rid)); err != nil {
-					// Should not block the main flow, just logged
+		if v, ok := change.Metadata.Parameters[sca.JobParameterRobot]; ok {
+			if jsonData, y := v.(string); y {
+				r := &model.Robot{}
+				if err := r.FromJSON(jsonData); err != nil {
 					log.Error(errors.Wrap(err, "scan controller: handle job hook"))
-				} else {
-					log.Debugf("Robot account with id %d for the scan %s is removed", int64(rid), trackID)
 				}
+
+				if r.ID > 0 {
+					if err := robot.RobotCtr.DeleteRobotAccount(r.ID); err != nil {
+						// Should not block the main flow, just logged
+						log.Error(errors.Wrap(err, "scan controller: handle job hook"))
+					} else {
+						log.Debugf("Robot account with id %d for the scan %s is removed", r.ID, trackID)
+					}
+				}
+
 			}
 		}
 	}
@@ -409,19 +417,14 @@ func (bc *basicController) GetStats(requester string) (*all.Stats, error) {
 	return sts, nil
 }
 
-// makeBasicAuthorization creates authorization from a robot account based on the arguments for scanning.
-func (bc *basicController) makeBasicAuthorization(pid int64, repository string, ttl int64) (string, int64, error) {
+func (bc *basicController) createRobotAccount(projectID int64, repository string) (*model.Robot, error) {
 	// Use uuid as name to avoid duplicated entries.
 	UUID, err := bc.uuid()
 	if err != nil {
-		return "", -1, errors.Wrap(err, "scan controller: make robot account")
+		return nil, errors.Wrap(err, "make robot account")
 	}
 
-	expireAt := time.Now().UTC().Add(time.Duration(ttl) * time.Second).Unix()
-
-	logger.Warningf("repository %s and expire time %d are not supported by robot controller", repository, expireAt)
-
-	resource := rbac.NewProjectNamespace(pid).Resource(rbac.ResourceRepository)
+	resource := rbac.NewProjectNamespace(projectID).Resource(rbac.ResourceRepository)
 	access := []*rbac.Policy{{
 		Resource: resource,
 		Action:   rbac.ActionScannerPull,
@@ -430,19 +433,17 @@ func (bc *basicController) makeBasicAuthorization(pid int64, repository string, 
 	robotReq := &model.RobotCreate{
 		Name:        UUID,
 		Description: "for scan",
-		ProjectID:   pid,
+		ProjectID:   projectID,
 		Access:      access,
+		Visible:     false,
 	}
 
 	rb, err := bc.rc.CreateRobotAccount(robotReq)
 	if err != nil {
-		return "", -1, errors.Wrap(err, "scan controller: make robot account")
+		return nil, errors.Wrap(err, "make robot account")
 	}
 
-	basic := fmt.Sprintf("%s:%s", rb.Name, rb.Token)
-	encoded := base64.StdEncoding.EncodeToString([]byte(basic))
-
-	return fmt.Sprintf("Basic %s", encoded), rb.ID, nil
+	return rb, nil
 }
 
 // launchScanJob launches a job to run scan
@@ -459,18 +460,7 @@ func (bc *basicController) launchScanJob(trackID string, artifact *v1.Artifact, 
 		return "", errors.Wrap(err, "scan controller: launch scan job")
 	}
 
-	var (
-		authorization string
-		rID           int64 = -1
-	)
-
-	if auth == authorizationBearer {
-		authorization, err = makeBearerAuthorization(artifact.Repository, fmt.Sprintf("%s:%s", registration.Name, registration.UUID))
-	} else {
-		// Make authorization from a robot account with 30 minutes
-		authorization, rID, err = bc.makeBasicAuthorization(artifact.NamespaceID, artifact.Repository, 1800)
-	}
-
+	robot, err := bc.createRobotAccount(artifact.NamespaceID, artifact.Repository)
 	if err != nil {
 		return "", errors.Wrap(err, "scan controller: launch scan job")
 	}
@@ -478,8 +468,7 @@ func (bc *basicController) launchScanJob(trackID string, artifact *v1.Artifact, 
 	// Set job parameters
 	scanReq := &v1.ScanRequest{
 		Registry: &v1.Registry{
-			URL:           registryAddr,
-			Authorization: authorization,
+			URL: registryAddr,
 		},
 		Artifact: artifact,
 	}
@@ -494,13 +483,17 @@ func (bc *basicController) launchScanJob(trackID string, artifact *v1.Artifact, 
 		return "", errors.Wrap(err, "launch scan job")
 	}
 
+	robotJSON, err := robot.ToJSON()
+	if err != nil {
+		return "", errors.Wrap(err, "launch scan job")
+	}
+
 	params := make(map[string]interface{})
 	params[sca.JobParamRegistration] = rJSON
 	params[sca.JobParameterRequest] = sJSON
 	params[sca.JobParameterMimes] = mimes
-	if rID > 0 {
-		params[sca.JobParameterRobotID] = rID
-	}
+	params[sca.JobParameterAuthType] = auth
+	params[sca.JobParameterRobot] = robotJSON
 
 	// Launch job
 	callbackURL, err := bc.config(configCoreInternalAddr)
@@ -519,24 +512,6 @@ func (bc *basicController) launchScanJob(trackID string, artifact *v1.Artifact, 
 	}
 
 	return bc.jc().SubmitJob(j)
-}
-
-// makeBearerAuthorization make a authorization with bearer token
-func makeBearerAuthorization(repository string, username string) (string, error) {
-	access := []*tk.ResourceActions{
-		{
-			Type:    "repository",
-			Name:    repository,
-			Actions: []string{rbac.ActionPull.String(), rbac.ActionScannerPull.String()},
-		},
-	}
-
-	accessToken, err := token.MakeToken(username, token.Registry, access)
-	if err != nil {
-		return "", errors.Wrap(err, "make bearer authorization")
-	}
-
-	return fmt.Sprintf("Bearer %s", accessToken.Token), nil
 }
 
 func parseOptions(options ...Option) (*Options, error) {

--- a/src/pkg/scan/api/scan/base_controller_test.go
+++ b/src/pkg/scan/api/scan/base_controller_test.go
@@ -15,7 +15,6 @@
 package scan
 
 import (
-	"encoding/base64"
 	"encoding/json"
 	"fmt"
 	"testing"
@@ -184,8 +183,7 @@ func (suite *ControllerTestSuite) SetupSuite() {
 	// Set job parameters
 	req := &v1.ScanRequest{
 		Registry: &v1.Registry{
-			URL:           "https://core.com",
-			Authorization: "Basic " + base64.StdEncoding.EncodeToString([]byte(common.RobotPrefix+"the-uuid-123:robot-account")),
+			URL: "https://core.com",
 		},
 		Artifact: suite.artifact,
 	}
@@ -196,12 +194,17 @@ func (suite *ControllerTestSuite) SetupSuite() {
 	regJSON, err := suite.registration.ToJSON()
 	require.NoError(suite.T(), err)
 
+	rb, _ := rc.CreateRobotAccount(account)
+	robotJSON, err := rb.ToJSON()
+	require.NoError(suite.T(), err)
+
 	jc := &MockJobServiceClient{}
 	params := make(map[string]interface{})
 	params[sca.JobParamRegistration] = regJSON
 	params[sca.JobParameterRequest] = rJSON
 	params[sca.JobParameterMimes] = []string{v1.MimeTypeNativeReport}
-	params[sca.JobParameterRobotID] = int64(1)
+	params[sca.JobParameterAuthType] = "Basic"
+	params[sca.JobParameterRobot] = robotJSON
 
 	j := &jm.JobData{
 		Name: job.ImageScanJob,

--- a/src/pkg/scan/job.go
+++ b/src/pkg/scan/job.go
@@ -16,14 +16,20 @@ package scan
 
 import (
 	"bytes"
+	"encoding/base64"
 	"encoding/json"
 	"fmt"
 	"reflect"
 	"sync"
 	"time"
 
+	"github.com/docker/distribution/registry/auth/token"
+	"github.com/goharbor/harbor/src/common"
+	"github.com/goharbor/harbor/src/common/config"
+	"github.com/goharbor/harbor/src/common/utils/registry/auth"
 	"github.com/goharbor/harbor/src/jobservice/job"
 	"github.com/goharbor/harbor/src/jobservice/logger"
+	"github.com/goharbor/harbor/src/pkg/robot/model"
 	"github.com/goharbor/harbor/src/pkg/scan/dao/scanner"
 	"github.com/goharbor/harbor/src/pkg/scan/report"
 	v1 "github.com/goharbor/harbor/src/pkg/scan/rest/v1"
@@ -37,11 +43,16 @@ const (
 	JobParameterRequest = "scanRequest"
 	// JobParameterMimes ...
 	JobParameterMimes = "mimeTypes"
-	// JobParameterRobotID ...
-	JobParameterRobotID = "robotID"
+	// JobParameterAuthType ...
+	JobParameterAuthType = "authType"
+	// JobParameterRobot ...
+	JobParameterRobot = "robotAccount"
 
 	checkTimeout       = 30 * time.Minute
 	firstCheckInterval = 2 * time.Second
+
+	authorizationBearer = "Bearer"
+	authorizationBasic  = "Basic"
 )
 
 // CheckInReport defines model for checking in the scan report with specified mime.
@@ -103,9 +114,18 @@ func (j *Job) Validate(params job.Parameters) error {
 		return errors.Wrap(err, "job validate")
 	}
 
-	// No need to check param robotID which os treated as an optional one.
-	// It is used to clear the generated robot account to reduce dirty data.
-	// Failure of doing this will not influence the main flow.
+	if _, err := extractRobotAccount(params); err != nil {
+		return errors.Wrap(err, "job validate")
+	}
+
+	authType, err := extractAuthType(params)
+	if err != nil {
+		return errors.Wrap(err, "job validate")
+	}
+
+	if authType != authorizationBearer && authType != authorizationBasic {
+		return errors.Wrapf(err, "job validate: not support auth type %s", authType)
+	}
 
 	return nil
 }
@@ -133,6 +153,25 @@ func (j *Job) Run(ctx job.Context, params job.Parameters) error {
 
 	// Ignore the namespace ID here
 	req.Artifact.NamespaceID = 0
+
+	robotAccount, _ := extractRobotAccount(params)
+
+	var authorization string
+	authType, _ := extractAuthType(params)
+	if authType == authorizationBearer {
+		tokenURL, err := getInternalTokenServiceEndpoint(ctx)
+		if err != nil {
+			return errors.Wrap(err, "scan job: get token service endpoint")
+		}
+		authorization, err = makeBearerAuthorization(robotAccount, tokenURL, req.Artifact.Repository)
+	} else {
+		authorization, err = makeBasicAuthorization(robotAccount)
+	}
+	if err != nil {
+		logAndWrapError(myLogger, err, "scan job: make authorization")
+	}
+
+	req.Registry.Authorization = authorization
 	resp, err := client.SubmitScan(req)
 	if err != nil {
 		return logAndWrapError(myLogger, err, "scan job: submit scan request")
@@ -331,6 +370,29 @@ func extractRegistration(params job.Parameters) (*scanner.Registration, error) {
 	return r, nil
 }
 
+func extractRobotAccount(params job.Parameters) (*model.Robot, error) {
+	v, ok := params[JobParameterRobot]
+	if !ok {
+		return nil, errors.Errorf("missing job parameter '%s'", JobParameterRobot)
+	}
+
+	jsonData, ok := v.(string)
+	if !ok {
+		return nil, errors.Errorf(
+			"malformed job parameter '%s', expecting string but got %s",
+			JobParameterRobot,
+			reflect.TypeOf(v).String(),
+		)
+	}
+	r := &model.Robot{}
+
+	if err := r.FromJSON(jsonData); err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
 func extractMimeTypes(params job.Parameters) ([]string, error) {
 	v, ok := params[JobParameterMimes]
 	if !ok {
@@ -357,4 +419,61 @@ func extractMimeTypes(params job.Parameters) ([]string, error) {
 	}
 
 	return mimes, nil
+}
+
+func extractAuthType(params job.Parameters) (string, error) {
+	v, ok := params[JobParameterAuthType]
+	if !ok {
+		return "", errors.Errorf("missing job parameter '%s'", JobParameterAuthType)
+	}
+
+	authType, ok := v.(string)
+	if !ok {
+		return "", errors.Errorf(
+			"malformed job parameter '%s', expecting string but got %s",
+			JobParameterAuthType,
+			reflect.TypeOf(v).String(),
+		)
+	}
+
+	return authType, nil
+}
+
+func getInternalTokenServiceEndpoint(ctx job.Context) (string, error) {
+	cfgMgr, ok := config.FromContext(ctx.SystemContext())
+	if !ok {
+		return "", errors.Errorf("failed to get config manager")
+	}
+
+	return cfgMgr.Get(common.CoreURL).GetString() + "/service/token", nil
+}
+
+// makeBasicAuthorization creates authorization from a robot account based on the arguments for scanning.
+func makeBasicAuthorization(robotAccount *model.Robot) (string, error) {
+	basic := fmt.Sprintf("%s:%s", robotAccount.Name, robotAccount.Token)
+	encoded := base64.StdEncoding.EncodeToString([]byte(basic))
+
+	return fmt.Sprintf("Basic %s", encoded), nil
+}
+
+// makeBearerAuthorization creates bearer token from a robot account
+func makeBearerAuthorization(robotAccount *model.Robot, tokenURL string, repository string) (string, error) {
+	cred := auth.NewBasicAuthCredential(robotAccount.Name, robotAccount.Token)
+	scopes := []*token.ResourceActions{
+		{
+			Type:    "repository",
+			Name:    repository,
+			Actions: []string{"pull"},
+		},
+	}
+	t, err := auth.GetToken(tokenURL, false, cred, scopes)
+	if err != nil {
+		return "", errors.Wrap(err, "scan job: make bearer token")
+	}
+
+	if t.GetToken() == "" {
+		return "", errors.New("scan job: bearer token is empty string")
+	}
+
+	return fmt.Sprintf("Bearer %s", t.GetToken()), nil
 }

--- a/src/pkg/scan/job_test.go
+++ b/src/pkg/scan/job_test.go
@@ -22,6 +22,7 @@ import (
 
 	"github.com/goharbor/harbor/src/jobservice/job"
 	"github.com/goharbor/harbor/src/jobservice/logger"
+	"github.com/goharbor/harbor/src/pkg/robot/model"
 	"github.com/goharbor/harbor/src/pkg/scan/dao/scanner"
 	v1 "github.com/goharbor/harbor/src/pkg/scan/rest/v1"
 	"github.com/goharbor/harbor/src/pkg/scan/vuln"
@@ -77,7 +78,7 @@ func (suite *JobTestSuite) TestJob() {
 	sr := &v1.ScanRequest{
 		Registry: &v1.Registry{
 			URL:           "http://localhost:5000",
-			Authorization: "the_token",
+			Authorization: "Basic cm9ib3Q6dG9rZW4=",
 		},
 		Artifact: &v1.Artifact{
 			Repository: "library/test_job",
@@ -89,12 +90,23 @@ func (suite *JobTestSuite) TestJob() {
 	sData, err := sr.ToJSON()
 	require.NoError(suite.T(), err)
 
+	robot := &model.Robot{
+		ID:    1,
+		Name:  "robot",
+		Token: "token",
+	}
+
+	robotData, err := robot.ToJSON()
+	require.NoError(suite.T(), err)
+
 	mimeTypes := []string{v1.MimeTypeNativeReport}
 
 	jp := make(job.Parameters)
 	jp[JobParamRegistration] = rData
 	jp[JobParameterRequest] = sData
 	jp[JobParameterMimes] = mimeTypes
+	jp[JobParameterAuthType] = "Basic"
+	jp[JobParameterRobot] = robotData
 
 	mc := &MockClient{}
 	sre := &v1.ScanResponse{

--- a/src/pkg/scan/rest/v1/models.go
+++ b/src/pkg/scan/rest/v1/models.go
@@ -123,8 +123,7 @@ func (s *ScanRequest) ToJSON() (string, error) {
 // Validate ScanRequest
 func (s *ScanRequest) Validate() error {
 	if s.Registry == nil ||
-		len(s.Registry.URL) == 0 ||
-		len(s.Registry.Authorization) == 0 {
+		len(s.Registry.URL) == 0 {
 		return errors.New("scan request: invalid registry")
 	}
 

